File: azure-network-topology-reviewer/engine/go/internal/analyze/analyze.go
// Package analyze is the deterministic analysis core — a direct port of
// reference/analyze.py. No model in the path; the same inputs always produce the
// same findings. The reference Python proves these functions against the fixtures.
package analyze

import (
	"fmt"
	"net/netip"
	"sort"
	"strings"

	"github.com/aaraminds/azure-nettopo-engine/internal/graph"
)

type Finding struct {
	Type      string `json:"type"`
	Severity  string `json:"severity"`
	Resource  string `json:"resource"`
	Evidence  string `json:"evidence"`
	Reachable bool   `json:"reachable"`
}

func cidrOverlap(a, b string) bool {
	pa, err1 := netip.ParsePrefix(a)
	pb, err2 := netip.ParsePrefix(b)
	if err1 != nil || err2 != nil {
		return false
	}
	return pa.Overlaps(pb)
}

func isInternetSource(s string) bool {
	switch strings.ToLower(s) {
	case "0.0.0.0/0", "internet", "*":
		return true
	}
	return false
}

func isBroadTagSource(s string) bool { return strings.EqualFold(s, "azurecloud") }

func nicVnet(n graph.NIC) string {
	if i := strings.Index(n.Subnet, "/"); i >= 0 {
		return n.Subnet[:i]
	}
	return ""
}

func contains(ss []string, s string) bool {
	for _, x := range ss {
		if x == s {
			return true
		}
	}
	return false
}

func deref(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

// adminVerdict — Gate 1. Highest-priority inbound AVNM verdict governing an
// internet-sourced flow on port. Source-scope aware: an Internet-tag rule does
// not govern intra-VNet/peered sources.
func adminVerdict(rules []graph.AdminRule, vnet, port string) string {
	best := ""
	bestPri := 1 << 30
	for _, ar := range rules {
		if ar.Direction != "Inbound" || !contains(ar.AppliesTo, vnet) || ar.DestinationPortRange != port {
			continue
		}
		src := strings.ToLower(ar.SourceAddressPrefix)
		if src != "internet" && src != "0.0.0.0/0" && src != "*" {
			continue
		}
		if ar.Priority < bestPri {
			bestPri, best = ar.Priority, ar.Access
		}
	}
	return best
}

// Analyze computes the findings for a topology. Deterministic.
func Analyze(fx *graph.Fixture) []Finding {
	var findings []Finding
	rg := fx.ResourceGraph
	effRules := fx.NetworkWatcher.EffectiveSecurityRules
	effRoutes := fx.NetworkWatcher.EffectiveRoutes
	admin := fx.AVNM.SecurityAdminRules
	fw := fx.AzureFirewall

	nics := map[string]graph.NIC{}
	for _, n := range rg.NetworkInterfaces {
		nics[n.Name] = n
	}

	for name, nic := range nics {
		rules := effRules[name]
		routes := effRoutes[name]
		hasPIP := nic.PublicIP != nil && *nic.PublicIP != ""
		defaultHop := ""
		for _, r := range routes {
			if r.AddressPrefix == "0.0.0.0/0" {
				defaultHop = r.NextHopType
				break
			}
		}
		vnet := nicVnet(nic)

		for _, r := range rules {
			if r.Direction != "Inbound" {
				continue
			}
			src := r.SourceAddressPrefix
			broadNet, broadTag := isInternetSource(src), isBroadTagSource(src)
			if !broadNet && !broadTag {
				continue
			}
			port := r.DestinationPortRange
			av := adminVerdict(admin, vnet, port) // Gate 1
			openInternet := r.Access == "Allow"   // Gate 2
			switch av {
			case "AlwaysAllow":
				openInternet = true
			case "Deny":
				openInternet = false
			}
			reachable := openInternet && hasPIP && defaultHop == "Internet" // Gates 3+4

			if reachable {
				sev := "High"
				if strings.EqualFold(nic.Tags["sensitive"], "true") {
					sev = "Critical"
				}
				ev := fmt.Sprintf("%s:%s inbound + route 0.0.0.0/0->Internet + public IP %s", src, port, deref(nic.PublicIP))
				if av == "AlwaysAllow" {
					ev += " (AVNM AlwaysAllow overrides NSG)"
				}
				if broadTag {
					ev += " — AzureCloud tag = all Azure public IPs, cross-tenant"
				}
				findings = append(findings, Finding{"over-permissive NSG (reachable)", sev, name, ev, true})
			} else {
				var why []string
				if !hasPIP {
					why = append(why, "no public IP")
				}
				if defaultHop == "None" {
					why = append(why, "route 0.0.0.0/0->None (black-hole)")
				} else if defaultHop != "" && defaultHop != "Internet" {
					why = append(why, "route 0.0.0.0/0->"+defaultHop)
				}
				if av == "Deny" {
					why = append(why, "AVNM Deny closes the Internet source (east-west may remain open)")
				}
				reason := strings.Join(why, "; ")
				if reason == "" {
					reason = "not reachable"
				}
				findings = append(findings, Finding{"over-permissive NSG (latent)", "Informational", name,
					fmt.Sprintf("%s:%s inbound but %s", src, port, reason), false})
			}
		}

		if fw != nil {
			for _, nat := range fw.NatRules {
				if nat.TranslatedAddress == nic.PrivateIP {
					findings = append(findings, Finding{"over-permissive NSG (reachable)", "High", name,
						fmt.Sprintf("firewall DNAT %s:%d -> %s:%d (source %v); no public IP on the NIC",
							fw.PublicIP, nat.DestinationPort, nic.PrivateIP, nat.TranslatedPort, nat.SourceAddresses), true})
				}
			}
		}
	}

	for _, pip := range rg.PublicIPAddresses {
		if pip.IPConfiguration == nil || *pip.IPConfiguration == "" {
			findings = append(findings, Finding{"orphaned public endpoint", "Low", pip.Name,
				fmt.Sprintf("public IP %s with null ipConfiguration", pip.IPAddress), false})
		}
	}

	vnets := rg.VirtualNetworks
	for i := 0; i < len(vnets); i++ {
		for j := i + 1; j < len(vnets); j++ {
			for _, pa := range vnets[i].AddressSpace {
				for _, pb := range vnets[j].AddressSpace {
					if cidrOverlap(pa, pb) {
						findings = append(findings, Finding{"CIDR overlap", "Medium",
							vnets[i].Name + "~" + vnets[j].Name,
							fmt.Sprintf("overlapping address space %s / %s", pa, pb), false})
					}
				}
			}
		}
	}

	// segmentation: sensitive subnet reachable VNet-wide via the default AllowVnetInBound
	for name, nic := range nics {
		if !strings.EqualFold(nic.Tags["sensitive"], "true") {
			continue
		}
		allowVnet, denyVnet := false, false
		for _, r := range effRules[name] {
			if r.Name == "AllowVnetInBound" || (r.Priority == 65000 && r.Access == "Allow") {
				allowVnet = true
			}
			if strings.Contains(r.Name, "DenyVnetInBound") {
				denyVnet = true
			}
		}
		if allowVnet && !denyVnet {
			findings = append(findings, Finding{"missing tier segmentation", "High", name,
				"sensitive subnet reachable VNet-wide via default AllowVnetInBound (no DenyVnetInBound above priority 65000)", true})
		}
	}

	sort.Slice(findings, func(i, j int) bool {
		if findings[i].Resource != findings[j].Resource {
			return findings[i].Resource < findings[j].Resource
		}
		return findings[i].Type < findings[j].Type
	})
	return findings
}

File: antr/engine/reference/analyze.py
#!/usr/bin/env python3
"""azure-nettopo-engine — deterministic analysis core (reference implementation).

This is the executable spec / test oracle for the production Go engine described in
NetworkTopologyReviewer-engine-plan.md. It is stdlib-only (no deps) and computes network
exposure deterministically from a topology export — no model in the path. The Go engine
(internal/analyze) is a direct port of these functions.

Run:  python analyze.py <fixture.json>   -> prints findings as JSON
"""
import ipaddress
import json
import sys


# ---------------------------------------------------------------- model / helpers
def cidr_overlap(a: str, b: str) -> bool:
    try:
        return ipaddress.ip_network(a, strict=False).overlaps(ipaddress.ip_network(b, strict=False))
    except ValueError:
        return False


def is_internet_source(src: str) -> bool:
    return (src or "").lower() in ("0.0.0.0/0", "internet", "*")


def is_broad_tag_source(src: str) -> bool:
    # service tags broader than they look (cross-tenant) — not raw internet, but over-permissive
    return (src or "").lower() in ("azurecloud",)


def finding(ftype, severity, resource, evidence, reachable):
    return {"type": ftype, "severity": severity, "resource": resource,
            "evidence": evidence, "reachable": reachable}


def _go_list(xs):
    # Format like Go fmt %v ([a b c]) so DNAT evidence matches the Go twin (V4-07).
    return "[" + " ".join(str(x) for x in (xs or [])) + "]"


def rid(obj) -> str:
    """Stable identity for a resource: its ARM resource id when present, else its
    bare name. Bare names are NOT unique across subscriptions/resource groups, so
    multi-subscription estates must carry `id`. (V4-07 — keeps single-sub fixtures
    that have no `id` byte-identical.)"""
    return obj.get("id") or obj.get("name", "")


def nic_vnet(nic) -> str:
    s = nic.get("subnet", "")
    return s.split("/")[0] if "/" in s else ""


# ---------------------------------------------------------------- Gate 1: AVNM admin rules
def admin_verdict(admin_rules, vnet, port, want_source="internet"):
    """Highest-priority inbound admin verdict that governs an `want_source`-sourced flow on `port`.
    Source-scope aware: an Internet-tag rule does NOT govern intra-VNet/peered sources."""
    best, best_pri = None, 10 ** 9
    for ar in admin_rules:
        if ar.get("direction") != "Inbound":
            continue
        if vnet not in (ar.get("appliesTo") or []):
            continue
        if str(ar.get("destinationPortRange")) != str(port):
            continue
        ars = (ar.get("sourceAddressPrefix") or "").lower()
        # only rules whose source covers the source under test apply
        if want_source == "internet" and ars not in ("internet", "0.0.0.0/0", "*"):
            continue
        pri = ar.get("priority", 10 ** 9)
        if pri < best_pri:
            best, best_pri = ar.get("access"), pri
    return best  # 'Allow' | 'AlwaysAllow' | 'Deny' | None


# ---------------------------------------------------------------- the analysis
def analyze(fx):
    rg = fx.get("resourceGraph", {})
    nw = fx.get("networkWatcher", {})
    admin_rules = fx.get("avnm", {}).get("securityAdminRules", [])
    fw = fx.get("azureFirewall")
    eff_rules = nw.get("effectiveSecurityRules", {})
    eff_routes = nw.get("effectiveRoutes", {})
    # Key NICs by stable identity (ARM id when present, else name). Keying by bare
    # name dropped same-named NICs across subscriptions at the INPUT stage (V4-07).
    nics = {rid(n): n for n in rg.get("networkInterfaces", [])}
    findings = []

    # Surface NICs whose Network Watcher enrichment failed (audit M-3).
    for _name in nw.get("incompleteNics", []):
        findings.append(finding("analysis incomplete", "Medium", _name,
                                "Network Watcher enrichment failed \u2014 effective rules/routes unavailable; NIC not evaluated for internet exposure", False))

    def eff_for(nic, table):
        # Network Watcher tables are keyed by NIC id in a real multi-sub estate;
        # current single-sub fixtures key by name. Try id, fall back to name.
        v = table.get(rid(nic))
        return v if v is not None else table.get(nic.get("name", ""), [])

    # ---- per-NIC internet exposure (Gates: AVNM source-scope -> NSG -> route -> public IP) ----
    for key, nic in nics.items():
        rules = eff_for(nic, eff_rules)
        routes = eff_for(nic, eff_routes)
        has_pip = bool(nic.get("publicIp"))
        default_hop = next((r.get("nextHopType") for r in routes if r.get("addressPrefix") == "0.0.0.0/0"), None)
        vnet = nic_vnet(nic)

        for r in rules:
            if r.get("direction") != "Inbound":
                continue
            src = r.get("sourceAddressPrefix", "")
            broad_net, broad_tag = is_internet_source(src), is_broad_tag_source(src)
            if not (broad_net or broad_tag):
                continue
            port = r.get("destinationPortRange", "")
            admin = admin_verdict(admin_rules, vnet, port)            # Gate 1
            nsg_allows = (r.get("access") == "Allow")                 # Gate 2 (effective)
            if admin == "AlwaysAllow":
                open_internet = True                                 # admin force-opens past NSG
            elif admin == "Deny":
                open_internet = False                                # admin closes the internet source
            else:
                open_internet = nsg_allows
            reachable = open_internet and has_pip and default_hop == "Internet"  # Gates 3+4

            if reachable:
                sensitive = str(nic.get("tags", {}).get("sensitive", "")).lower() == "true"
                sev = "Critical" if sensitive else "High"
                ev = f"{src}:{port} inbound + route 0.0.0.0/0->Internet + public IP {nic.get('publicIp')}"
                if admin == "AlwaysAllow":
                    ev += " (AVNM AlwaysAllow overrides NSG)"
                if broad_tag:
                    ev += " — AzureCloud tag = all Azure public IPs, cross-tenant"
                findings.append(finding("over-permissive NSG (reachable)", sev, rid(nic), ev, True))
            else:
                why = []
                if not has_pip:
                    why.append("no public IP")
                if default_hop == "None":
                    why.append("route 0.0.0.0/0->None (black-hole)")
                elif default_hop and default_hop != "Internet":
                    why.append(f"route 0.0.0.0/0->{default_hop}")
                if admin == "Deny":
                    why.append("AVNM Deny closes the Internet source (east-west may remain open)")
                # NOTE: parenthesize the `or` — `+` binds tighter, so the fallback
                # "not reachable" was previously dead code (LOW-1).
                findings.append(finding("over-permissive NSG (latent)", "Informational", rid(nic),
                                        f"{src}:{port} inbound but " + ("; ".join(why) or "not reachable"), False))

        # inbound firewall DNAT publishes a no-public-IP backend
        if fw:
            for nat in fw.get("natRules", []):
                if nat.get("translatedAddress") == nic.get("privateIp"):
                    findings.append(finding(
                        "over-permissive NSG (reachable)", "High", rid(nic),
                        f"firewall DNAT {fw.get('publicIp')}:{nat.get('destinationPort')} -> "
                        f"{nic.get('privateIp')}:{nat.get('translatedPort')} (source {_go_list(nat.get('sourceAddresses'))}); "
                        f"no public IP on the NIC", True))

    # ---- orphaned public endpoints ----
    for pip in rg.get("publicIPAddresses", []):
        if not pip.get("ipConfiguration"):
            findings.append(finding("orphaned public endpoint", "Low", rid(pip),
                                    f"public IP {pip.get('ipAddress')} with null ipConfiguration", False))

    # ---- CIDR / address-space overlap ----
    vnets = rg.get("virtualNetworks", [])
    for i in range(len(vnets)):
        for j in range(i + 1, len(vnets)):
            for pa in vnets[i].get("addressSpace", []):
                for pb in vnets[j].get("addressSpace", []):
                    if cidr_overlap(pa, pb):
                        findings.append(finding("CIDR overlap", "Medium",
                                                f"{vnets[i]['name']}~{vnets[j]['name']}",
                                                f"overlapping address space {pa} / {pb}", False))

    # ---- segmentation: sensitive subnet reachable VNet-wide via the default AllowVnetInBound ----
    for key, nic in nics.items():
        if str(nic.get("tags", {}).get("sensitive", "")).lower() != "true":
            continue
        rules = eff_for(nic, eff_rules)
        allow_vnet = any(r.get("name") == "AllowVnetInBound" or
                         (r.get("priority") == 65000 and r.get("access") == "Allow") for r in rules)
        deny_vnet = any("DenyVnetInBound" in (r.get("name", "")) for r in rules)
        if allow_vnet and not deny_vnet:
            findings.append(finding("missing tier segmentation", "High", rid(nic),
                                    "sensitive subnet reachable VNet-wide via default AllowVnetInBound "
                                    "(no DenyVnetInBound above priority 65000)", True))

    # Canonical, deterministic ordering — must match the Go engine's sort
    # (internal/analyze/analyze.go: by resource, then type) so the reference and
    # the port emit findings in the same order and stay true twins.
    # Total order incl. evidence — a NIC can emit two findings with the same
    # (resource, type) (e.g. two latent rules); evidence breaks the tie so the
    # order is fully determined and matches the Go engine's SliceStable comparator.
    findings.sort(key=lambda f: (f["resource"], f["type"], f["evidence"]))
    return findings


def main():
    if len(sys.argv) != 2:
        sys.exit("usage: python analyze.py <fixture.json>")
    fx = json.load(open(sys.argv[1], encoding="utf-8"))
    print(json.dumps(analyze(fx), indent=2))


if __name__ == "__main__":
    main()
